src/System.IO.Pipelines/BufferSegment.cs

```diff
@@ -85,11 +85,11 @@ public BufferSegment(OwnedMemory<byte> buffer, int start, int end)
 
         public void Dispose()
         {
-            Debug.Assert(_buffer.ReferenceCount >= 1);
+            Debug.Assert(!_buffer.HasZeroReferences);
```

Comment:
I think `DisposableReference` and  ` HasOutstandingReferences` makes the most sense.  As `AddReference` exists on the API surface as well. 

Would `AutoReference` be better than `DisposableReference`?